il.cpp

```diff
@@ -677,6 +678,10 @@ bool GetLowLevelILForInstruction(Architecture* arch, uint64_t addr, LowLevelILFu
 	InstructionOperand& operand3 = instr.operands[2];
 	InstructionOperand& operand4 = instr.operands[3];
 
+	// these opcodes can't occur once aarch64_decompose() has been called
```

Comment:
I believe this to be true because I assembled ubfm and sbfm with every possible encoding of operands 3 and 4 (immediate arguments in range 0-64), and disassembled the resulting bin with radare2 and binja. [us]bfm was _never_ preferred over their aliases (LSL, LSR, UBFIZ, UBFX, UXTB, UXTH)

XREF http://shell-storm.org/armv8-a/ISA_v85A_A64_xml_00bet8/xhtml/ubfm.html

See also BFXPreferred() in the ARM docs which seems to catch-all for UBFX and SBFX.

Reply:
Is it preferred to leave the assert in, or to log a warning?

fwiw ARM64_UBFM and ARM64_UBFX are not the same operation so leaving ARM64_SBFM in the switch case will produce incorrectly lifted IL. From the [docs](http://shell-storm.org/armv8-a/ISA_v85A_A64_xml_00bet8/xhtml/ubfx_ubfm.html):

```
UBFX <Xd>, <Xn>, #<lsb>, #<width>
```

is equivalent to

```
UBFM <Xd>, <Xn>, #<lsb>, #(<lsb>+<width>-1)
```

Reply:
You are right: `UBFM_32M_BITFIELD` and `UBFM_64M_BITFIELD` are encodings that are never reachable because they are completely masked by aliases. There are other cases where an alias encoding is specified to never be preferred, so it just sits in the spec, stale. I've counted 52 of these strange cases total: ENC_ASRV_32_DP_2SRC, ENC_ASRV_64_DP_2SRC, ENC_BFM_32M_BITFIELD, ENC_BFM_64M_BITFIELD, ENC_BIC_AND_Z_ZI_, ENC_CMPLE_CMPGE_P_P_ZZ_, ENC_CMPLO_CMPHI_P_P_ZZ_, ENC_CMPLS_CMPHS_P_P_ZZ_, ENC_CMPLT_CMPGT_P_P_ZZ_, ENC_DUP_ASISDONE_ONLY, ENC_EON_EOR_Z_ZI_, ENC_FACLE_FACGE_P_P_ZZ_, ENC_FACLT_FACGT_P_P_ZZ_, ENC_FCMLE_FCMGE_P_P_ZZ_, ENC_FCMLT_FCMGT_P_P_ZZ_, ENC_FMOV_CPY_Z_P_I_, ENC_FMOV_DUP_Z_I_, ENC_HINT_HM_HINTS, ENC_INS_ASIMDINS_IV_V, ENC_INS_ASIMDINS_IR_R, ENC_LDRB_32BL_LDST_REGOFF, ENC_LDRSB_32BL_LDST_REGOFF, ENC_LDRSB_64BL_LDST_REGOFF, ENC_LDR_BL_LDST_REGOFF, ENC_LSLV_32_DP_2SRC, ENC_LSLV_64_DP_2SRC, ENC_LSRV_32_DP_2SRC, ENC_LSRV_64_DP_2SRC, ENC_MOVN_32_MOVEWIDE, ENC_NOT_ASIMDMISC_R, ENC_NOT_EOR_P_P_PP_Z, ENC_ORN_ORR_Z_ZI_, ENC_REV64_REV_64_DP_1SRC, ENC_RORV_32_DP_2SRC, ENC_RORV_64_DP_2SRC, ENC_SBFM_32M_BITFIELD, ENC_SBFM_64M_BITFIELD, ENC_STRB_32BL_LDST_REGOFF, ENC_STR_BL_LDST_REGOFF, ENC_UBFM_32M_BITFIELD, ENC_UBFM_64M_BITFIELD, ENC_UDF_ONLY_PERM_UNDEF, ENC_UMOV_ASIMDINS_X_X, ENC_CPY_Z_O_I_, ENC_CPY_Z_P_I_, ENC_CPY_Z_P_R_, ENC_CPY_Z_P_V_, ENC_DUP_Z_I_, ENC_DUP_Z_R_, ENC_DUPM_Z_I_, ENC_FCPY_Z_P_I_, ENC_FDUP_Z_I_

Reply:
> Is it preferred to leave the assert in, or to log a warning?

My initial thought, subject to change, is to have no check at all, or at most a warning. I don't like the assert. What's the worst that can happen? The disassembly doesn't correctly alias `UBFM_32M_BITFIELD` or `SBFM_64M_BITFIELD` and it gets lifted? This seems the most flexible option.

Reply:
Agreed, done.